remill/Arch/Runtime/Operators.h

```diff
@@ -465,6 +465,73 @@ MAKE_WRITE_REF(float64_t)
 
 #undef MAKE_WRITE_REF
 
+#define MAKE_CMPXCHG(size, type_prefix, access_suffix) \
+    template <typename T> \
+    ALWAYS_INLINE static bool _CmpXchg( \
+        Memory *&memory, RnW<T> op, type_prefix ## size ## _t &expected, \
+        type_prefix ## size ## _t desired) { \
+        auto prev_val = expected; \
+        if (*op.val_ref == expected) {\
+          expected = *reinterpret_cast<type_prefix ## size ## _t*>(op.val_ref); \
+          *op.val_ref = desired; \
+        } \
+        return prev_val == expected; \
+    } \
+    \
+    template <typename T> \
+    ALWAYS_INLINE static bool _CmpXchg( \
+        Memory *&memory, MnW<T> op, type_prefix ## size ## _t &expected, \
+        type_prefix ## size ## _t desired) { \
+        auto prev_val = expected; \
+        memory = __remill_compare_exchange_memory_ ## access_suffix (memory, op.addr, expected, desired);\
+        return prev_val == expected; \
+    }
+
+MAKE_CMPXCHG(8, uint, 8)
+MAKE_CMPXCHG(16, uint, 16)
+MAKE_CMPXCHG(32, uint, 32)
+MAKE_CMPXCHG(64, uint, 64)
+MAKE_CMPXCHG(128, uint, 128)
+
+#undef MAKE_CMPXCHG
+#define UCmpXchg(op, oldval, newval) _CmpXchg(memory, op, oldval, newval)
+
+#define MAKE_ATOMIC_INTRINSIC(name, size, type_prefix, op)\
+  template<typename T> \
+  ALWAYS_INLINE type_prefix ## size ## _t _ ## name(Memory *&memory, MnW<T> addr, type_prefix ## size ## _t &value) { \
+  memory = __remill_ ## name ## _ ## size(memory, addr.addr, value); \
+    return value; \
+  } \
+  \
+  template<typename T> \
+  ALWAYS_INLINE type_prefix ## size ## _t _ ## name(Memory *&memory, RnW<T> addr, type_prefix ## size ## _t &value) { \
+    auto prev_value = *reinterpret_cast<type_prefix ## size ## _t*>(addr.val_ref); \
+    *addr.val_ref = prev_value op value; \
+    value = prev_value; \
+    return value; \
+  } \
+
+#define MAKE_ATOMIC(name, op, make_atomic_intrinsic)  \
+    make_atomic_intrinsic(name, 8, uint, op) \
+    make_atomic_intrinsic(name, 16, uint, op) \
+    make_atomic_intrinsic(name, 32, uint, op) \
+    make_atomic_intrinsic(name, 64, uint, op) \
+
+MAKE_ATOMIC(fetch_and_add, +, MAKE_ATOMIC_INTRINSIC)
+MAKE_ATOMIC(fetch_and_sub, -, MAKE_ATOMIC_INTRINSIC)
+MAKE_ATOMIC(fetch_and_or, |, MAKE_ATOMIC_INTRINSIC)
+MAKE_ATOMIC(fetch_and_and, &, MAKE_ATOMIC_INTRINSIC)
+MAKE_ATOMIC(fetch_and_xor, ^, MAKE_ATOMIC_INTRINSIC)
+//MAKE_ATOMIC(fetch_and_nand, MAKE_ATOMIC_INTRINSIC)
+
+#define fetch_and_add(op1, op2) _fetch_and_add(memory, op1, op2)
```

Comment:
Can you make these names follow the naming convention? E.g. `UFetchAdd16`, for example.